### ISSUE TITLE
Invalidate cache on config change

### ISSUE THREAD

**Issue by LoicGrobol** [flakeheaven/flakeheaven#71]
This one has puzzled me for hours: I had `max_line_length = 100` in my `pyproject.toml`, yet flakeheaven was reporting lines of length between 100 and 90 as too long. After a shamingly long time spend in `pdb` to try to understand why, I figured it out: I first ran `flakeheaven` in this repo without configuration, so it found these lines too long by the default configuration, and even after I changed the config, since the files were not changed, I was still getting the old reports, even if they shouldn't have been errors anymore.

TL;DR currently if a file has cached results, these results are still used even if the config has changed since then, which can lead to wrong reports.

Side note: the cache system is a cool feature in some cases, but I would really like to be able to disable it with a more convenient way than setting `FLAKEHEAVEN_CACHE_TIMEOUT` to 0.

PS: thanks for this cool project!

**Comment by pwoolvett** [comment]
what about adding the ` tool.flakeheaven` sections contents to the cache  hash ?

**Comment by LoicGrobol** [comment]
@pwoolvett Would that be enough? What about the other config sources?

**Comment by pwoolvett** [comment]
> @pwoolvett Would that be enough? What about the other config sources?

You're right. It's not as easy as I first tought, as the "final" configuration options could include (potentially) remote http refs, local `.toml`s, `pyproject.toml`, and cli...

maybe we should look at how other systems impement/invalidate cache. Out of my mid, mypy / pylint seem they work fine.

**Comment by pwoolvett** [comment]
@LoicGrobol 

Can you confirm your issue happened only with configuration in the main `[tool.flakeheaven]` section? Specifically, I think everything inside `[tool.flakeheaven.plugins]` should be working with the cache. Ege see [here](https://github.com/flakeheaven/flakeheaven/blob/03157d0e673a929508f2627eed5b4dcc7313beec/flakeheaven/_logic/_snapshot.py#L38-L52)
